consensus/src/txn_hash_and_authenticator_deduper.rs

```diff
@@ -0,0 +1,344 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    counters::{TXN_DEDUP_FILTERED, TXN_DEDUP_SECONDS},
+    transaction_deduper::TransactionDeduper,
+};
+use aptos_types::transaction::SignedTransaction;
+use rayon::prelude::*;
+use std::collections::{HashMap, HashSet};
+
+pub struct TxnHashAndAuthenticatorDeduper {}
+
+impl TransactionDeduper for TxnHashAndAuthenticatorDeduper {
+    fn dedup(&self, transactions: Vec<SignedTransaction>) -> Vec<SignedTransaction> {
+        let _timer = TXN_DEDUP_SECONDS.start_timer();
+        let mut seen = HashMap::new();
+        let mut is_possible_duplicate = false;
+        let mut possible_duplicates = vec![false; transactions.len()];
+        for (i, txn) in transactions.iter().enumerate() {
+            match seen.get(&(txn.sender(), txn.sequence_number())) {
+                None => {
+                    seen.insert((txn.sender(), txn.sequence_number()), i);
+                },
+                Some(first_index) => {
+                    is_possible_duplicate = true;
+                    possible_duplicates[*first_index] = true;
```

Comment:
I think it is possible to bucket the (hash, authenticator) comparisons with minimal effort, i.e. set possible_duplicates[i] to point to first_index, like finding connected components, except we know the root :)

Then, if and when it's needed, we could easily parallelize the comparing the hashes and signatures per component (maybe in the future we will get access to the cached hash from earlier stage and that becomes a bottleneck), by assigning a dedup data-structure to each first_index (component) and either having a thread handle a whole component (so no extra synchronization), or more granular per txn by locking the corresponding dedup data-structure - most likely earlier is better to avoid locking overhead despite a bit worse load balancing.
Anyway, maybe we don't parallelize now, but we could still identify the components and adding a comment to the later part for dedup (will note there).

Dedup per component / bucket is safe since there is no way to have the same hash if sender / seq num aren't the same.